include/zephyr/usb/class/usb_midi.h

```diff
@@ -0,0 +1,149 @@
+ * @see ump112: "Universal MIDI Packet (UMP) Format and MIDI 2.0 Protocol"
+ *              Document version 1.1.2
+ * @{
+ */
+
+#include <zephyr/device.h>
+
+/**
+ * @defgroup usb_midi_mt Universal MIDI Packet message types
+ * @ingroup usb_midi
+ * @see ump112: 2.1.4 Message Type (MT) Allocation
+ * @{
+ */
+
+#define MT_UTILITY             0x00
+#define MT_SYS_RT_COMMON       0x01
+#define MT_MIDI1_CHANNEL_VOICE 0x02
+#define MT_DATA                0x03
+#define MT_MIDI2_CHANNEL_VOICE 0x04
+#define MT_FLEX_DATA           0x0d
+#define MT_UMP_STREAM          0x0f
+
+/**
+ * @}
+ *
+ * @defgroup usb_midi_accessors Universal MIDI Packet accessors
+ * @ingroup usb_midi
+ * @see ump112: 2.1.4 Message Type (MT) Allocation
+ * @{
+ *
+ * @brief      Size of a Universal MIDI Packet, in 32bit words
+ * @param[in]  ump    The universal MIDI Packet
+ * @see ump112: 2.1.4 Message Type (MT) Allocation
+ */
+#define UMP_WORDS(ump) (1 + ((0xfe950d40 >> (2 * UMP_MT(ump))) & 3))
+
+/**
+ * @brief      Message Type field of a Universal MIDI Packet
+ * @param[in]  ump    The universal MIDI Packet
+ */
+#define UMP_MT(ump)    ((ump)[0] >> 28)
+/**
+ * @brief      MIDI group field of a Universal MIDI Packet
+ * @param[in]  ump    The universal MIDI Packet
+ */
+#define UMP_GROUP(ump) (((ump)[0] >> 24) & 0x0f)
+
+/**
+ * @brief      Command of a MIDI1 channel voice message
+ * @param[in]  ump    A universal MIDI Packet (containing a MIDI1 event)
+ */
+#define UMP_MIDI1_COMMAND(ump) (((ump)[0] >> 20) & 0x0f)
+/**
+ * @brief      Channel of a MIDI1 channel voice message
+ * @param[in]  ump    A universal MIDI Packet (containing a MIDI1 event)
+ */
+#define UMP_MIDI1_CHANNEL(ump) (((ump)[0] >> 16) & 0x0f)
+/**
+ * @brief      First parameter of a MIDI1 channel voice message
+ * @param[in]  ump    A universal MIDI Packet (containing a MIDI1 event)
+ */
+#define UMP_MIDI1_P1(ump)      (((ump)[0] >> 8) & 0x7f)
+/**
+ * @brief      Second parameter of a MIDI1 channel voice message
+ * @param[in]  ump    A universal MIDI Packet (containing a MIDI1 event)
+ */
+#define UMP_MIDI1_P2(ump)      ((ump)[0] & 0x7f)
+
+/**
+ * @}
+ *
+ * @brief      Initialize a MIDI1 Universal Midi Packet
+ * @param      group    The UMP group
+ * @param      command  The MIDI1 command nibble
+ * @param      channel  The MIDI1 channel number
+ * @param      p1       The 1st MIDI1 parameter
+ * @param      p2       The 2nd MIDI1 parameter
+ */
+#define UMP_MIDI1(group, command, channel, p1, p2)                                                 \
+	(const uint32_t[])                                                                         \
+	{                                                                                          \
+		(MT_MIDI1_CHANNEL_VOICE << 28) | ((group & 0x0f) << 24) |                          \
+			((command & 0x0f) << 20) | ((channel & 0x0f) << 16) | ((p1 & 0x7f) << 8) | \
+			(p2 & 0x7f)                                                                \
+	}
```

Comment:
Well, this was my first proposal, but @tmon-nordic suggested to use macros instead, see discussion https://github.com/zephyrproject-rtos/zephyr/pull/81197#discussion_r1839755320 .

Having written both versions (with structs / macros), I have a personal preference for macros. This is because structs require an additional typecast for proper interpretation, for example:

```C
uint32_t packet[4];

if (UMP_MT(packet) == UMP_MT_MIDI1_CHANNEL_VOICE) {
    struct ump_midi1_channel_voice *cv = packet;
    // For example NoteOn / 4 bytes
} else if (UMP_MT(ump) == UMP_MT_DATA) {
    struct ump_data *data = packet;
    // For example SysEx messages / 8 bytes
} else if (UMP_MT(ump) == UMP_MT_SYS_RT_COMMON) {
    struct ump_sys_rt_common *sysrt = packet;
    // For example MIDI clock / 4 bytes
} else if (UMP_MT(ump) == UMP_MT_UMP_STREAM) {
    struct ump_stream *stream = packet;
    // For example MIDI Capability inquiry / 16 bytes
}
```

We may `union` them into a nice top-level `ump` struct, but I feel this is getting complicated, and would yield a bunch of repeated code.

Moreover, some accessor macros can be used for different message types. For example, consider the MIDI1 Channel Voice and System messages that have a similar structure:

![image](https://github.com/user-attachments/assets/3bd20c6d-b63f-4a89-a520-5e8322eb599f)
![image](https://github.com/user-attachments/assets/0696c7c0-20aa-46b8-bdc3-e1741b864e32)

Reply:
What I suggested is widely used in Zephyr, e.g. in Bluetooth and networking. You can find many good examples there.
But packet processing in usbd_midi.c should be limited to what is described in "3.2 Data format": Universal MIDI Packet (UMP)", anything else UMP specific should be placed in midi.h (you would need to include in the sample), and maybe in the future there will be a MIDI support that can also be used by e.g. Bluetooth.
As I understand midi class spec, all you need to know before the driver callback is called is the "Message Type". For this you can use a macro defined in usbd_midi.c, if necessary. And then pull the message from the net_buf and pass it to the user driver using a callback like `typedef void (*usb_midi_callback)(const struct device *dev, const uint32_t msg[static 4]);`

Reply:
I splitted the definitions between `midi.h` (Universal MIDI Packet definitions) and `usbd_midi.h` and took the opportunity to improve the doxygen comments. Also changed typing from `const uint32_t *ump` to `const uint32_t ump[static 4]`.

About the macros vs. structs I'm still not convinced (especially considering the number of cases if we were to add more definitions for MIDI2.0 messages; see ump112 section 7.4). Let me know how you think it looks currently; I will rewrite to structs if necessary.

> and maybe in the future there will be a MIDI support that can also be used by e.g. Bluetooth.

Yes, would be awesome to have MIDI support in other "connectivity" stacks :star_struck: